cli/progress.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+use std::sync::Arc;
+use std::sync::Mutex;
+
+#[derive(Clone, Default)]
+pub struct Progress(Arc<Mutex<Inner>>);
+
+impl Progress {
+  pub fn new() -> Self {
+    Progress::default()
+  }
+
+  pub fn set_callback<F>(&self, f: F)
+  where
+    F: Fn(bool, usize, usize, &str) + Send + Sync + 'static,
+  {
+    let mut s = self.0.lock().unwrap();
+    assert!(s.callback.is_none());
+    s.callback = Some(Arc::new(f));
+  }
+
+  /// Returns job counts: (complete, total)
+  pub fn progress(&self) -> (usize, usize) {
+    let s = self.0.lock().unwrap();
+    s.progress()
+  }
+
+  pub fn history(&self) -> Vec<String> {
+    let s = self.0.lock().unwrap();
+    s.job_names.clone()
+  }
+
+  pub fn add(&self, name: String) -> Job {
+    let mut s = self.0.lock().unwrap();
+    let id = s.job_names.len();
+    s.maybe_call_callback(false, s.complete, s.job_names.len() + 1, &name);
+    s.job_names.push(name);
+    Job {
+      id,
+      inner: self.0.clone(),
+    }
+  }
+
+  pub fn done(&self) {
+    let s = self.0.lock().unwrap();
+    s.maybe_call_callback(true, s.complete, s.job_names.len(), "");
+  }
+}
+
+type Callback = dyn Fn(bool, usize, usize, &str) + Send + Sync;
+
+#[derive(Default)]
+struct Inner {
+  job_names: Vec<String>,
+  complete: usize,
+  callback: Option<Arc<Callback>>,
+}
+
+impl Inner {
+  pub fn maybe_call_callback(
+    &self,
+    done: bool,
+    complete: usize,
+    total: usize,
+    msg: &str,
+  ) {
+    if let Some(ref cb) = self.callback {
+      cb(done, complete, total, msg);
+    }
+  }
+
+  /// Returns job counts: (complete, total)
+  pub fn progress(&self) -> (usize, usize) {
+    let total = self.job_names.len();
+    (self.complete, total)
+  }
+}
+
+pub struct Job {
+  inner: Arc<Mutex<Inner>>,
+  id: usize,
+}
+
+impl Drop for Job {
+  fn drop(&mut self) {
+    let mut s = self.inner.lock().unwrap();
+    s.complete += 1;
+    let name = &s.job_names[self.id];
+    let (complete, total) = s.progress();
+    s.maybe_call_callback(false, complete, total, name);
```

Comment:
I would expect the `done` parameter as passed to the callback to be 'true' here.
After all having all files loaded is not the only condition in which a job could end, e.g it could also have errored out.
It'd be cleaner to notify the callback that "this is the last time you're getting called", regardless of whether the job was actually 100% successful.

Reply:
I don't agree - it's a single "job" (download/compilation), there's no way to know if it was the last one, that's why I suggested explicit `Progress.done` method.

Reply:
@bartlomieju is correct, "done" is for when the entire progress is done, this is just a single job.